programs/stake/src/stake_state.rs

```diff
@@ -842,6 +848,69 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
     }
 }
 
+#[derive(Copy, Clone, Debug, Default)]
+struct ValidatedSplitInfo {
+    source_remaining_balance: u64,
+    dest_rent_exempt_reserve: u64,
+}
+
+fn validate_split_amount(
+    source: &KeyedAccount,
+    dest: &KeyedAccount,
+    lamports: u64,
+) -> Result<Option<ValidatedSplitInfo>, InstructionError> {
+    let source_lamports = source.lamports()?;
+
+    // verify enough account lamports
+    if lamports > source_lamports {
+        return Err(InstructionError::InsufficientFunds);
+    }
+
+    match source.state()? {
+        StakeState::Stake(meta, _) | StakeState::Initialized(meta) => {
+            // verify that the source account still has enough lamports left
+            // EITHER at least the minimum balance, OR zero (in this case the source
+            // account is transferring all lamports to new destination account, and the source
+            // account will be closed)
+            let source_minimum_balance = meta
+                .rent_exempt_reserve
+                .saturating_add(MINIMUM_STAKE_DELEGATION);
+            let source_remaining_balance = source_lamports.saturating_sub(lamports);
+            if source_remaining_balance == 0 {
+                // full amount is a withdrawal
+                // nothing to do here
+            } else if source_remaining_balance < source_minimum_balance {
+                // the remaining balance is too low to do the split
+                return Err(InstructionError::InsufficientFunds);
```

Comment:
(i wonder we should start to use StakeDelegationTooSmall or not for consistent behavior...)

Reply:
I thought about this too. There are multiple error cases:

1. Split amount leaves src account below rent exempt minimum
2. Split amount leaves dst account below rent exempt minimum
3. Split amount leaves src account below (rent exempt minimum + minimum stake delegation)
4. Split amount leaves dst account below (rent exempt minimum + minimum stake delegation)

What should each error be?

(1) and (2) I think are straight forward to leave as erroring with InsufficientFunds. (3) and (4) are more up in the air. I ended up settling on the idea that the split amount _does_ have insufficient funds to satisfy both rent exempt minimum _and_ the minimum stake delegation, which is how the code works today. I agree it may be beneficial to have further context returned in the error to know if the insufficient funds were due to not meeting rent exempt minimum, or if it was for not meeting rent exempt minimum + minimum stake delegation.

Since this PR is not changing current behavior, maybe this can turn into its own separate issue and PR to resolve it. Wdyt about separate issue? Also, wdyt about which errors to use?

Reply:
wow quite good writeup for summarizing the current state regarding this. thanks

so, next question is why did we introduce `StakeDelegationTooSmall` in this pr in the first place? can't we just reuse (i.e. = overload...) the `InsufficientFunds`?

don't take me wrong. i love context-sensitive detailed errors. so, i like `StakeDelegationTooSmall` than the too generic `InsufficientFunds`. however, avoiding consistency is more important, in my opinion. this benefits to developers for learning and ensuring security with consistent reasoning

conceptually, trying to creat a new stake account should functionally be equivalent to splitting full amount from a existing stake account. so ideally same error should be returned for both cases if lamports are too small.

currently, both code path results in `InsufficientFunds`. however, starting from this pr, the creating case starts to return `StakeDelegationTooSmall`, right? I wonder we rather want to stick to `InsufficientFunds` for now. and later address the context-sensitive errors if wanted as a separate issue?

Reply:
Yes, I can do that. I'll get working on that ASAP and also open a new issue to track adding the StakeTooSmall error.

Reply:
Added a task in https://github.com/solana-labs/solana/issues/22559 to track adding the StakeDelegationTooSmall error.